File: proyectoRuby/lib/mazo_sorpresas.rb
module Civitas
class MazoSorpresas
   
  private
  def init
    @sorpresas = []
    @cartasEspeciales = []
    @barajada = false
    @usadas = 0
  end
  
  public
  def initialize(b=false)
    @debug = b
    init
    
    if @debug
      Diario.instance.ocurre_evento("Modo debug del mazo de sorpresas activado")
    end
  end
  
  def al_mazo(s)
    if !@barajada
      @sorpresas.push(s)
    end
  end
  
  def siguiente
    if (!@barajada || @usadas == @sorpresas.length) && !@debug
      @sorpresas.shuffle
      @usadas = 0
      @barajada = true
    end
    
    @usadas += 1
    @ultimaSorpresa = @sorpresas[0]
    @sorpresas.delete_at(0)
    @ultimaSorpresa
  end
  
  def inhabilitar_carta_especial(s)
    if @sorpresas.include?(s)
      @sorpresas.delete_at( @sorpresas.index(s) )
      @cartasEspeciales.push(s)
      Diario.instance.ocurre_evento("Carta especial " + s.nombre + " deshabilitada")
    end
  end
  
  def habilitar_carta_especial(s)
    if @cartasEspeciales.include?(s)
      @cartasEspeciales.delete_at( @cartasEspeciales.index(s) )
      @sorpresas.push(s)
      Diario.instance.ocurre_evento("Carta especial " + s.nombre + " habilitada")
    end
  end
end
end

File: proyectoJava/src/civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Random;
import java.util.Collections;

public class MazoSorpresas {
    // Atributos
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    // Métodos
    private void init (){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(){
        init();
        debug = false;
    }
    
    MazoSorpresas(boolean b){
        debug = b;
        init();
        
        if (debug){
            Diario.getInstance().ocurreEvento("Se ha activado el modo debug del MazoSorpresas");
        }
    }
    
    void alMazo (Sorpresa s){
        if (!barajada)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente (){
        if ( (!barajada || usadas == sorpresas.size() ) && !debug){
            Collections.shuffle(sorpresas, new Random());
            usadas = 0;
            barajada = true;
        }
            
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
        
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial (Sorpresa sorpresa){
        int i = sorpresas.indexOf(sorpresa);
        if (i != -1){
            sorpresas.remove(i);
            cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Carta especial deshabilitada");
        };
    }
    
    void habilitarCartaEspecial (Sorpresa sorpresa){
        int i = cartasEspeciales.indexOf(sorpresa);
        
        if (i != -1){
            cartasEspeciales.remove(i);
            sorpresas.add(sorpresa);
            Diario.getInstance().ocurreEvento("Carta especial habilitada");
        };
    }
}
